### ISSUE TITLE
Missing SocketTaskExtensions from System.Net.Sockets on Desktop

### ISSUE THREAD

**Issue by safern** [dotnet/corefx#17690]
This is breaking developers and tests. 

We should consider fixing the tests in the short-term.

Should we fix this to bring this type back to desktop?

cc: @davidsh

**Comment by davidsh** [comment]
cc: @weshaggard @karelz @stephentoub @CIPop

**Comment by davidsh** [comment]
This is due to the change in packaging where we no longer build a separate System.Net.Sockets NuGet package.  Thus, we no longer bundle the partial facade for System.Net.Sockets.  That partial facade has the SocketTaskExtensions class defined.

**Comment by weshaggard** [comment]
We are still having a conversation about exactly how we want to handle these. At this point we are leaning towards shipping a .NET Framework Support package which contains these types of libraries but we don't have full buy-off yet. For now we should temporarily disable the tests or enable the build of this library for netfx that can enable the tests.

**Comment by davidsh** [comment]
>For now we should temporarily disable the tests or enable the build of this library for netfx that can enable the tests.

We can't disable the tests because this extension class is used as part of implementing the loopback test server.  So, it impacts a ton of tests running on Desktop for Sockets, Security, Http, etc.

My plan was to just copy the code and/or create a common file for the `SocketTaskExtensions` class.  Then include the file in the built of the tests to unblock the Desktop test runs.

**Comment by weshaggard** [comment]
>My plan was to just copy the code and/or create a common file for the SocketTaskExtensions class. Then include the file in the build of the tests to unblock the Desktop test runs.

That sounds reasonable to me as well.

**Comment by karelz** [comment]
Sounds good, thanks for the unblocking plan @davidsh!

@weshaggard do we have a bug tracking finalizing the plan in 2.0 timeframe?

**Comment by karelz** [comment]
cc @danmosemsft

**Comment by weshaggard** [comment]
I've filed https://github.com/dotnet/standard/issues/274.

**Comment by karelz** [comment]
Great, thanks!

**Comment by davidsh** [comment]
>My plan was to just copy the code and/or create a common file for the SocketTaskExtensions class. Then include the file in the build of the tests to unblock the Desktop test runs.

So, my plan is not going to work easily. I started looking at this.  It is not just the LoopBackServer.cs that uses the Task extensions for Socket.  But other files in Http tests use various *Async extension methods as well.

And to make matters worse, the compiler is binding those extension method calls to netstandard.dll!SocketTaskExtensions.  So, I can't just define the extension class somewhere else because we'll still get a TypeLoadException.  I thought about creating a special SocketEx subclass and defining the *Async overrides there but that is more work and too much test code across the repo to fix.

So, we really a way for netstandard.dll to properly include/reference SocketTaskExtension for System.Net.Sockets.  And currently those extension methods are calling into internal Socket methods to do the actual work. 

Until we get this solved, there will be lots of .NET Core tests that can't run due on Desktop to TypeLoadException.

**Comment by karelz** [comment]
Thanks @davidsh! In that case I propose to push it to post-2.0.
We will run a bit blind, but given all the work we have, this is lower priority. @danmosemsft thoughts?

**Comment by davidsh** [comment]
>Thanks @davidsh! In that case I propose to push it to post-2.0.

Not fixing this issue will block #17691 which is marked for 2.0.

**Comment by karelz** [comment]
So we push part of that one into Future as well.
We need to be practical and keep our real goal in mind -- the goal of this exercise is to ensure compatibility between Desktop and .NET Core. All under the assumption that running CoreFX tests on Desktop is simple. If it is costly for specific area to do, we may need to reconsider the value of that exercise vs. delivering product.

**Comment by danmosemsft** [comment]
I don't have full context, but I agree, the desktop effort is important but not essential in all areas for 2.0

**Comment by karelz** [comment]
Another angle how to look at this is:
Is there something else, tactical, lower-ish cost, even hacky or temporary, that we could do to ensure Networking area is compatible between Desktop and .NET Core? If not -> Future. If yes, let's discuss/consider it.

Overall it is the usual cost/benefit/risk assessment for a release.
@davidsh what is your recommendation based on all the info above?

**Comment by davidsh** [comment]
>Overall it is the usual cost/benefit/risk assessment for a release.
@davidsh what is your recommendation based on all the info above?

I think it is very valuable to be able to run the .NET Core test collateral against NETFX (Desktop).  We have already found good bugs to fix into terms of making .NET Core compatible with NETFX (See: #17701).

I have some new ideas I will be exploring this weekend to help unblock this. I will report back later.

**Comment by karelz** [comment]
Great, thanks!

**Comment by safern** [comment]
Thanks @davidsh I agree with you, it is very valuable to have tests running in netfx. We already have vendors looking at the failures and will start opening bugs which might be compat issues. So thanks for your effort to unblock this and have the tests running against netfx in the networking area.

**Comment by weshaggard** [comment]
>And to make matters worse, the compiler is binding those extension method calls to netstandard.dll!SocketTaskExtensions.

@davidsh to work around this just add a netfx specific configuration for the tests so it will build against the netfx targeting pack instead of the netstandard targeting pack. That should enable you to add the extensions. 

I also agree we should try to get the test running on desktop before 2.0 ships to help ensure we don't break anything.

**Comment by davidsh** [comment]
>@davidsh to work around this just add a netfx specific configuration for the tests so it will build against the netfx targeting pack instead of the netstandard targeting pack. That should enable you to add the extensions.

It turns out that this was more complicated than that.  The SocketTaskExtensions is listed as a type forward from the System.Net.Sockets facade (pure facade). That points to netstandard.dll.  But netstandard.dll is completing missing the SocketTaskExtensions type.  That is why the type load exception is being thrown.

I didn't want to ifdef the test code for NETFX compile but the majority of tests are being built against netstandard.  And for the most part, that is the way our customers will build things. I also tried to do runtime detection in the loopback server code such as:

```c#
if (PlatformDetection.IsFullFramework)
{
    // don't use SocketTaskExtensions
}
else
{
    // ok to use SocketTaskExtension
}
```

but that blows up with the same type load exception. The reason is that the JIT isn't happy because the type is being referenced in the source code and it can't find it, even though it won't be executed on Full Framework.

So, I decided on a simpler approach to working around this problem and got the Http tests fully building and running. See: #17775

**Comment by stephentoub** [comment]
> It turns out that this was more complicated than that

I just tried and believe I got this working using basically the described approach.  I'll clean it up and submit a PR.  22 of 947 tests in System.Net.Sockets.Tests are failing after the change.

**Comment by stephentoub** [comment]
@karelz, why did you assign this to me?

**Comment by stephentoub** [comment]
This is being addressed by https://github.com/dotnet/corefx/pull/18246 (and then if that doesn't work for some reason, by https://github.com/dotnet/corefx/pull/18278).

**Comment by CIPop** [comment]
Reopening since there are two cases where TODO items in the test-code are referencing this issue.

https://github.com/dotnet/corefx/search?utf8=%E2%9C%93&q=17690&type=
